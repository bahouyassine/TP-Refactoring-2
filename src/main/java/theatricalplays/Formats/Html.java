package theatricalplays.Formats;

import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import theatricalplays.File;
import theatricalplays.Invoice;
import theatricalplays.PerformanceInvoice;
import theatricalplays.Play;

public class Html {
    private static StringBuffer PerformanceInvoiceHtmlFormat(List<PerformanceInvoice> performanceInvoices){
        NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);
        StringBuffer result = new StringBuffer("");
        for (PerformanceInvoice perf_inv : performanceInvoices) {
            result.append(String.format(
            "<tr>\n"
            + "<td>%s</td>\n"
            + "<td>%s</td>\n"
            + "<td>%s</td>\n"
            + "</tr>\n" ,
            perf_inv.playName,
            perf_inv.audience,
            frmt.format(perf_inv.Amount)
            ));
        }
        
        return result;
        }
    public static String toHtml(Invoice invoice,Map<String, Play> plays) {
        String htmlFilePath =  "src/main/resources/template/TemplateHtml.html";
        String htmlTemplate = File.readFileToString(htmlFilePath);
        invoice.CalculateInvoice(plays);
        String performanceInvoiceHtml = PerformanceInvoiceHtmlFormat(invoice.performanceInvoices).toString();
        String[][] replacements = {
            {"{@Customer_Name}", invoice.customer.name},
            {"{@Performance_Invoice}", performanceInvoiceHtml},
            {"{@Invoice_Amount}", String.valueOf(invoice.totalPrice)},
            {"{@Total_Credits}", String.valueOf(invoice.credit)},
        };

        for (String[] replacement : replacements) {
            htmlTemplate = htmlTemplate.replace(replacement[0], replacement[1]);
        }

        return htmlTemplate;
    }




}
