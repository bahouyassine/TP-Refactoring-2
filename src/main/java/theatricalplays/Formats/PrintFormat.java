package theatricalplays.Formats;


public enum PrintFormat { 
    HTML,
    TXT
} 


