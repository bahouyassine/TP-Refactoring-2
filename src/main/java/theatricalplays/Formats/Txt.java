package theatricalplays.Formats;

import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import theatricalplays.File;
import theatricalplays.Invoice;
import theatricalplays.PerformanceInvoice;
import theatricalplays.Play;

public class Txt {

    private static StringBuffer PerformanceInvoiceTxtFormat(List<PerformanceInvoice> performanceInvoices){
        NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);
        StringBuffer result = new StringBuffer("");
        for (PerformanceInvoice perf_inv : performanceInvoices) {
            result.append(String.format(
            "  %s: %s (%s seats)\n" ,
            perf_inv.playName,
            frmt.format(perf_inv.Amount),
            perf_inv.audience
            ));
        }
        
        return result;
    }

    public static String toText(Invoice invoice,Map<String, Play> plays)  {
        NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);

        String filePath = "/home/local.isima.fr/yabahou/Desktop/TP-Refactoring-2/src/main/resources/template/TemplateTxt.txt";
        String template = File.readFileToString(filePath);
        invoice.CalculateInvoice(plays);
        String performanceInvoiceTxt = PerformanceInvoiceTxtFormat(invoice.performanceInvoices).toString();
        String[][] replacements = {
            {"{@Customer_Name}", invoice.customer.name+"\n"},
            {"{@Performance_Invoice}", performanceInvoiceTxt},
            {"{@Invoice_Amount}", frmt.format((invoice.totalPrice))+"\n"},
            {"{@Total_Credits}", String.valueOf(invoice.credit)}
        };
        for (String[] replacement : replacements) {
            template = template.replace(replacement[0], replacement[1]);
            }
        template = template +"\n" +DiscountToText(invoice.totalPrice,invoice.credit);
        return template;
    }

    public static String DiscountToText(double totalPrice,int credit)  {
        String template = "";
        if (Invoice.ApplyReduction(totalPrice,credit)){
            NumberFormat frmt = NumberFormat.getCurrencyInstance(Locale.US);
            String filePath = "/home/local.isima.fr/yabahou/Desktop/TP-Refactoring-2/src/main/resources/template/TemplateTxtDiscount.txt";
            template = File.readFileToString(filePath);
            String[][] replacements = {
                {"{@enter}","\n"}, 
                {"{@Invoice_Amount_after_discount}", frmt.format(totalPrice)},
                {"{@Total_Credits_after_discount}", String.valueOf(credit)}
            };
            for (String[] replacement : replacements) {
                template = template.replace(replacement[0], replacement[1]);
            }

        } 

        return template;
    }

}
