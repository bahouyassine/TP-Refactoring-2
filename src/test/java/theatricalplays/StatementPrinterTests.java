package theatricalplays;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

import theatricalplays.Formats.PrintFormat;

import java.util.List;
import java.util.Map;

import static org.approvaltests.Approvals.verify;

public class StatementPrinterTests {

    @Test
    void TxtTest() {
        Map<String, Play> plays = Map.of(
                "hamlet",  new Play("Hamlet", TheaterPlayType.TRAGEDY),
                "as-like", new Play("As You Like It", TheaterPlayType.COMEDY),
                "othello", new Play("Othello", TheaterPlayType.TRAGEDY));

        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)));

        StatementPrinter statementPrinter = new StatementPrinter();
        var result = statementPrinter.print(invoice, plays,PrintFormat.TXT);
        verify(result);
    }

    @Test
    void TxtTestWithDiscount() {
        Map<String, Play> plays = Map.of(
                "hamlet",  new Play("Hamlet", TheaterPlayType.TRAGEDY),
                "as-like", new Play("As You Like It", TheaterPlayType.COMEDY),
                "othello", new Play("Othello", TheaterPlayType.TRAGEDY));

        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance("hamlet", 100),
                new Performance("as-like", 120),
                new Performance("othello", 90)));

        StatementPrinter statementPrinter = new StatementPrinter();
        var result = statementPrinter.print(invoice, plays,PrintFormat.TXT);
        verify(result);
    }

    @Test
    void htmlTest() {
        Map<String, Play> plays = Map.of(
                "hamlet",  new Play("Hamlet", TheaterPlayType.TRAGEDY),
                "as-like", new Play("As You Like It", TheaterPlayType.COMEDY),
                "othello", new Play("Othello", TheaterPlayType.TRAGEDY));

        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance("hamlet", 55),
                new Performance("as-like", 35),
                new Performance("othello", 40)));

        StatementPrinter statementPrinter = new StatementPrinter();
        var result = statementPrinter.print(invoice, plays,PrintFormat.HTML);
        verify(result);
    }

    @Test
    void statementWithNewPlayTypes() {
        Map<String, Play> plays = Map.of(
                "henry-v",  new Play("Henry V", TheaterPlayType.HISTORY),
                "as-like", new Play("As You Like It", TheaterPlayType.PASTORAL));

        Invoice invoice = new Invoice("BigCo", List.of(
                new Performance("henry-v", 53),
                new Performance("as-like", 55)));
        StatementPrinter statementPrinter = new StatementPrinter();
        Assertions.assertThrows(Error.class, () -> {
            statementPrinter.print(invoice, plays,PrintFormat.TXT);

        });
    }
}
