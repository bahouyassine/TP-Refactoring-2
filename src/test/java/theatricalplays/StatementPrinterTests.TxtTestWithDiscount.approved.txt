Statement for BigCo
  Hamlet: $1,100.00 (100 seats)
  As You Like It: $1,260.00 (120 seats)
  Othello: $1,000.00 (90 seats)
Amount owed is $3,360.00
You earned 244 credits
Congratulations you earned a loyalty discount!!!
A loyalty discount of â‚¬15 has been applied.
Amount owed after discount: $3,360.00
Credits left after discount: 244